README

Lua Shelve Module r0.3 README
Adrian Perez <mobius@iespana.es>


------[ Introduction ]-----------------------------------------------

This is a Lua module that uses either GNU Gdbm or the more standard
Ndbm database  libraries to store  data in files,  but this data is
saved in a special way:  all standard Lua datatypes may be directly
stored and  retrieved in its original form and  contents, including
Lua tables. A shelf file is a special kind of userdata that behaves
similarly to  a Lua table, so its contents are  accessed in a table-
like fashion. Quick example:

	-- Create data
	my_table = {
		name = "This is a string within a table",
		level= 1,
		is_at_level_one = true,
		another_table = {
			name = "This is a nested table",
			level= 2,
			is_at_level_one = false
		}
	}
	
	-- Open a 'shelve' file
	file = shelve.open("data")
	
	-- This encodes & stores the above table
	-- Dotted and bracketed syntax may be used interchangeably
	file["my_table"] = my_table
	file.title = "File containing a string and a table"
	
	-- Show contents of the file in (key, value) form
	for key in file() do
		print("(" .. key .. ", " .. db[key] .. ")")
	end

The name of the module, "shelve", is taken from the Python's module
with the same name, as both work in a similar way. The nuts & bolts
of the module are coded directly in C, so operation speed is fairly
fast.


------[ Requirements ]-----------------------------------------------

In order to build and use the this Lua module, you will need the 
following:
	
	- Working C compiler, GCC recommended.
	- GNU DBM (aka Gdbm) or the standard Unix Ndbm. If you have the
		old Unix DBM please upgrade to Gdbm or Ndbm. Gdbm is better.
	- Lua 5.0 beta or above, including headers and libraries.
	- GNU Make (ok, you may also  build manually...).  I don't know
		wether other versions of Make will work.

If you want to compile the module as a dynamically loadable library
(plugin) you will also need the following:

	- Lua interpreter with "loadlib()" support. Most modern operating
		systems (Linux, MacOS X, Windows, BSD...) can handle this.


------[ Operation ]--------------------------------------------------

To gain access to a data file, you must open it first. The module
defined an "open()" function that receives two parameters:

	file_handle = shelve.open(file_name, access_mode)

	- file_name is the name of the file you want to open.

	- access_mode is a string describing how the file should be
		accessed.  "r" means readonly, and  any other string will
		is assumed to be read+write access. If you don't pass the
		second parameter, read+write access is assumed. Note that
		the mode in which a file is opened is important: a writer
		needs exclusive  access to the file,  but various readers
		may read it concurrently!  When you need only read-access,
		please use this option,  and you will get better response
		when several processes/threads  are working with the same
		data file.

In order to retain the structure and type information of the data,
it must be encoded with this  information included. The result of
coding is a stream of bytes. The 'shelve' data files are standard
Gdbm/Ndbm files  that contain these  encoded bytestreams  as data,
and the specified identifier as key of the data, so the following
Lua statement:

	file["a number"] = 12345
	
Will use the string "a number" as key and the encoded form of the
"12345" numeric value as data in the Gdbm/Ndbm file.

The opposite action:

	num = file["a number"]
	
Will look-up for the "a number" key in the Gdbm/Ndbm file. If the
key does NOT exist, a "nil" value is returned, otherwise the byte-
stream associated with the key  will be read and decoded in order
to return the original Lua structure.

To delete a (key , data) pair from  the file, just assign the key
to "nil". Example:

	file = shelve.open("test")
	file.number = 12345   -- this assignment defines "number"
	file.number = nil			-- this assignment deletes "number"
	
Indexing a shelve file with an undefined key will always return
"nil". Example:

	file = shelve.open("test")
	file.number = 12345   -- this assignment defines "number"
	print(file.number)		-- prints "12345"
	file.number = nil			-- this assignment deletes "number"
	print(file.number)		-- prints "nil"
	print(file.undefined)	-- prints "nil", again

In order to close a shelf,  just assign it to 'nil',  and the Lua
garbage collector will close the file in its next pass. When Gdbm
is used as backend  and the file was open in  read+write mode, it
will be reorganized on close.


------[ Extras ]-----------------------------------------------------

The module also defines two standalone functions that perform the
encoding/decoding of data:

	- shelve.marshal(data):
		Encodes "data" and returns the encoded bytestream as a Lua
		string.
	
	- shelve.unmarshal(bytestream):
		Takes an encoded "bytestream"  as a Lua string and returns 
		the Lua data stored after decoding it.
		
The relantionship between these two functions is that one is the
inverse of the other, so:

		data1 == shelve.unmarshal(shelve.marshal(data1)) => true
		data2 == shelve.marshal(shelve.unmarshal(data2)) => true

You may use these functions to  manually encode and decode data. 
Note that shelve.marshal() adds a special mark to the end of an
encoded bytestream:  decoding the concatenation  of two encoded
values will only return the first one.


------[ MT-Safety ]--------------------------------------------------

All the functions contained in this Lua module are reentrant, so it
is multithread-safe. There's no shared globals or static storage.

--------------------------------------------------------[ EOF ]------


ChangeLog

2003-04-23	Adrian Perez	<mobius@iespana.es>
	* shelve.c: removed stupid xfree(udata) in the __gc metamethod.

2003-04-17	Adrian Perez	<mobius@iespana.es>
	* defs.h: fixed strdup() replacement.
	* marshal.c: updated to use the Lua auxiliar library.
	* shelve.c: updated to use the Lua auxiliar library, the code is
			more readable now, specially in the Lua-interfacing functions.

2003-04-10	Adrian Perez	<mobius@iespana.es>
	* defs.h: added strdup() replacement.
	* shelve.c: metatable used in the 'shelf' userdata is no longer
			stored in the namespace table of the module, it's in the Lua
			registry instead.

2003-04-08	Adrian Perez	<mobius@iespana.es>
	* shelve.c: added '__tostring' metamethod to the 'shelf' userdata,
			the returned string contains file name and read-write mode info.
	* shelve.c: fixed some bugs derived from popping string values from
			the stack before using them.
	* shelve.c: fixed small memory leak (I forgot a free() in the '__gc' 
			metamethod).
	* ChangeLog: ChangeLog started.



test.lua

local printf
if unpack ~= nil then
   printf = function(fmt, ...)
      io.stdout:write(string.format(fmt, unpack(arg)))
   end
else
   printf = function(fmt, ...)
      io.stdout:write(string.format(fmt, ...))
   end
end

package.cpath = "./?.so"
local shelve = assert(require("shelve"))
print("Opening 'db'")

local db = assert(shelve.open("test.db"))

a = "one"
b = "two"
c = "three"

db.num = 1234567890 -- a number
db.str = "a string" -- a string
db.t1  = {}         -- an empty table
db.t2  = { s="S" }  -- table with one element
db.t3  = { a,b,c }  -- indexed table, multiple elements

db.nested =         -- nested tables
   {
      level = a,
      nest  = {
         level = b,
         nest  = {
            level = c,
            nest  = "nothing"
         }
      }
   }

printf("Number encoding... ")
assert(type(db.num) == "number")
assert(db.num == 1234567890)
printf("ok\nString encoding... ")
assert(db.str == "a string")
printf("ok\nTable encoding... ")
assert(type(db.t1) == "table")
assert(type(db.t2) == "table")
assert(type(db.t3) == "table")
printf("ok\nData integrity... ")
assert(db.t2.s == "S")
assert(db.t3[1] == a)
assert(db.t3[2] == b)
assert(db.t3[3] == c)
t = db.nested
assert(type(t) == "table")
assert(t.level == a)
assert(type(t.nest) == "table")
assert(t.nest.level == b)
assert(type(t.nest.nest) == "table")
assert(t.nest.nest.level == c)
assert(type(t.nest.nest.nest) == "string")
assert(t.nest.nest.nest == "nothing")
printf("ok\nLarge file storing... ")

fd, err = io.open("LICENSE", "r")
if (not fd) then
   print(err)
   os.exit()
end
lgpl_license = fd:read("*a")
fd:close()

db.lic = lgpl_license
assert(db.lic == lgpl_license)
printf("ok\n... all tests successfully passed ...\n")


defs.h
/*
 * Common definitions.
 * Copyright (C) 2003-2016 Adrian Perez de Castro <aperez@igalia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef __defs__h
#define __defs__h

/*
 * Header declarations start/end macros. These protect definitions
 * to be mangled by C++ compilers, so linking with plain C functions
 * and variables is done correctly.
 */
#if defined(__cplusplus) || defined(c_plusplus)
# define BEGIN_DECLS  extern "C" {
# define END_DECLS    }
#else
# define BEGIN_DECLS
# define END_DECLS
#endif


/*
 * Chain of fools to detect wether we are using Win32.
 */
#if defined(__WINDOWS__) || defined(_WINDOWS) || defined(_Windows) || \
    defined(__WIN32__)   || defined(_WIN32)   || defined(WIN32)    || \
    defined(__NT__) || defined(__NT_DLL__) || defined(__WINDOWS_386__)
# define PLAT_WIN32
#else
# define PLAT_POSIX
#endif


/*
 * The API macro is used to mark functions that must be exported to
 * the world. Under Win32 it's expanded to export symbols in DLLs
 * when appropiate.
 */
#if defined(PLAT_WIN32) && defined(BUILD_DLL)
# define API extern __declspec(dllexport)
#elif defined(PLAT_WIN32) && defined(USE_DLL)
# define API extern __declspec(dllimport)
#else
# define API extern
#endif


/*
 * Exits the program printing a string to stdout.
 */
extern void exit();
#include <stdio.h>
#define __die(__str) \
    (fprintf(stderr, "fatal: %s\n", __str), exit(-1))

/*
 * The xmalloc(), xrealloc() and xfree() act as expected, but they
 * do pointer checks when needed. They are defined as inlines or
 * macros, so using them does not involve extra function calls.
 * "__inline" is used to prevent complains of old silly compilers.
 */
extern void* malloc();
extern void* realloc();
extern void free();

#ifdef __GNUC__
# define INLINE inline
#else
# define INLINE __inline
#endif

static INLINE void* xmalloc(size_t sz)
{
    register void *p = malloc(sz);
    return (p ? p : (__die("out of memory"), (void*)0));
}

#define xrealloc(__p, __nsz) \
    ((__p) = realloc(__p, __nsz), (__p) ? __p : (__die("out of memory"), (void*)0))

#define xfree(__p) \
    (free(__p), (__p) = NULL)


/*
 * strdup() replacement
 */
#if !defined(HAVE_STRDUP) && (defined(_OSX) || defined(_LINUX))
# define HAVE_STRDUP
#endif

#if defined(HAVE_STRDUP) && defined(__GNUC__)
# define xstrdup(__strp) \
    ({ char *__$strp = strdup(__strp); \
       (__$strp) ? : (__die("out of memory"), (char*)0); })
#else
# ifdef __GNUC__
#  define xstrdup(__strp) \
    ({ const char* __$strp = __strp; \
       strcpy((char*) xmalloc(sizeof(char)*(strlen(__$strp)+1)), __$strp); })
# else
static INLINE char* xstrdup(const char *strp)
{
    return strcpy((char*) xmalloc(sizeof(char) * (strlen(strp) + 1)), strp);
}
# endif
#endif

#endif /* !__defs__h */


anydb.h
/*
 * Gdbm/Ndbm wrapper header.
 * Wrappers are modelled after the Gdbm prototypes.
 *
 * Copyright (C) 2003-2016 Adrian Perez de Castro <aperez@igalia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef __anydb__h
#define __anydb__h

/* Useable DBM backends */
#define ANYDB_NDBM  1
#define ANYDB_GDBM  2

/* Handle Autoconf like HAVE_GDBM and HAVE_NDBM, Gdbm used by default */
#ifndef ANYDB_BACKEND
# if defined(HAVE_GDBM)
#  define ANYDB_BACKEND  ANYDB_GDBM
# elif defined(HAVE_NDBM)
#  define ANYDB_BACKEND  ANYDB_NDBM
# else
#  define ANYDB_BACKEND  ANYDB_GDBM
# endif
#endif

/* Use rw-rw--- mode by default */
#ifndef ANYDB_OPEN_MODE
# define ANYDB_OPEN_MODE  0660
#endif

/* Define wrappers depending on the used backend */
#if (ANYDB_BACKEND == ANYDB_NDBM)
# include <ndbm.h>             /* Backend header file */
# include <fcntl.h>            /* O_CREAT and O_RDWR macros */
# define ANYDB_READ            O_RDONLY
# define ANYDB_WRITE           (O_CREAT | O_RDWR)
# define ANYDB_INSERT          DBM_INSERT
# define ANYDB_REPLACE         DBM_REPLACE
  typedef DBM* anydb_t;        /* Database handle type */
# define anydb_open(_n,_f)     dbm_open(_n, _f, ANYDB_OPEN_MODE)
# define anydb_close           dbm_close
# define anydb_store           dbm_store
# define anydb_fetch           dbm_fetch
# define anydb_delete          dbm_delete
# define anydb_firstkey        dbm_firstkey
# define anydb_nextkey(_h,_k)  dbm_nextkey(_h)
# define anydb_reorganize(_i)  ((void)0)
#elif (ANYDB_BACKEND == ANYDB_GDBM)
# include <gdbm.h>             /* Backend header file */
# define ANYDB_READ            GDBM_READER
# define ANYDB_WRITE           GDBM_WRCREAT
# define ANYDB_INSERT          GDBM_INSERT
# define ANYDB_REPLACE         GDBM_REPLACE
  typedef GDBM_FILE anydb_t;   /* Database handle type */
# define anydb_open(_n,_f)     gdbm_open(_n, 0, _f, ANYDB_OPEN_MODE, NULL)
# define anydb_close           gdbm_close
# define anydb_store           gdbm_store
# define anydb_fetch           gdbm_fetch
# define anydb_delete          gdbm_delete
# define anydb_firstkey        gdbm_firstkey
# define anydb_nextkey         gdbm_nextkey
# define anydb_reorganize      gdbm_reorganize
#else
  /* No backend defined -> error */
# error Neither Ndbm or Gdbm configured.
#endif

#endif /* !__anydb__h */


shelve.h
/*
 * Marshaling/Unmarshaling functions, Shelve module exports.
 * Copyright (C) 2003-2016 Adrian Perez de Castro
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */


#ifndef __lua_module_shelve__h
#define __lua_module_shelve__h

#include "defs.h"
#include <lua.h>

/*
 * Key used to store the metatable for "shelve" files in the
 * Lua registry.
 */
#define SHELVE_REGISTRY_KEY "shelve-file-meta"

/*
 * Lua to byte-array marshaling functions. Representation of
 * numbers is machine-dependant.
 *
 * Format:
 *   'n'XXXX   Number: XXXX=value, sizeof(lua_Number)
 *   's'XXXXS  String: XXXX=length, sizeof(size_t), S=contents.
 *   'b'       Boolean, false.
 *   'B'       Boolean, true.
 *   't'XXXX{} Table: XXXX=number of items, sizeof(size_t).
 *             Each element: KV.
 *    KV       K: key, either nXXXX or sXXXXS (only in tables).
 *             V: value, either nXXXX, sXXXXS, b, B or t{}.
 *    '!'      End-of-sequence marker.
 *    'T'      End-of-table marker.
 */

#define MARSH_NUMBER  'n'
#define MARSH_STRING  's'
#define MARSH_TRUE    'B'
#define MARSH_FALSE   'b'
#define MARSH_TABLE   't'
#define MARSH_EOT     'T'
#define MARSH_EOS     '!'

int shelve_marshal(lua_State*, char**, int*);
int shelve_unmarshal(lua_State*, const char**);
API int l_shelve_marshal(lua_State*);
API int l_shelve_unmarshal(lua_State*);
API int l_shelve_init(lua_State*);

#endif /* !__lua_module_shelve__h */


marshal.c

/*
 * (Un)Marshaling of Lua datatypes.
 * Copyright (C) 2003-2016 Adrian Perez de Castro <aperez@igalia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include <string.h>
#include <lua.h>
#include <lauxlib.h>
#include <assert.h>
#include "defs.h"
#include "shelve.h"


#define GROW(_dp, _szvar, _sz) \
    _szvar += (_sz); \
    _dp = ((_dp) ? xrealloc(_dp, _szvar * sizeof(char)) \
                 : xmalloc(_szvar * sizeof(char)))

#define STOR(_dp, _szvar, _what, _sz) \
    { GROW(_dp, _szvar, _sz); \
      memcpy(&(_dp)[(_szvar) - (_sz)], _what, _sz); }


int marshal_table(lua_State*, char**, int*);
int unmarshal_table(lua_State*, const char**);


int
shelve_unmarshal(lua_State *L, const char **datap)
{
    size_t ssz;
    const char *data = *datap;

    assert(L);
    assert(data);

    switch (*(data++)) {
        case MARSH_FALSE: /* 'false' boolean value */
            lua_pushboolean(L, 0);
            break;
        case MARSH_TRUE: /* 'true' boolean value */
            lua_pushboolean(L, 1);
            break;
        case MARSH_NUMBER: /* numeric value */
            lua_pushnumber(L, (lua_Number) (*((double*) data)) );
            data += sizeof(double);
            break;
        case MARSH_STRING: /* string value */
            ssz   = *((size_t*) data);
            data += sizeof(size_t);
            lua_pushlstring(L, data, ssz);
            data += ssz;
            break;
        case MARSH_TABLE: /* table */
            if (!unmarshal_table(L, &data)) return 0;
            break;
        default:
            return 0;
    }

    if (*(data++) != MARSH_EOS) return 0;

    *datap = data;
    return 1;
}


int
unmarshal_table(lua_State *L, const char **datap)
{
    const char *data = *datap;

    assert(L);
    assert(data);

    lua_newtable(L);
    /*
     * stack: tbl
     * index:  -1
     */

    for (;;) {
        switch (*data) {
            case MARSH_NUMBER: /* valid key (number) */
            case MARSH_STRING: /* valid key (string) */
                if (!shelve_unmarshal(L, &data) ||
                    !shelve_unmarshal(L, &data)) return 0;
                /*
                 * stack: data key tbl
                 * index:   -1  -2  -3
                 */
            lua_rawset(L, -3);
            break;
            case MARSH_EOT: /* end of table */
            data++;
            /*
             * I know using 'goto' is not structured programming, but
             * I use it to avoid cheching "*data" two times. Without
             * 'goto' the code would look like the following:
             *
             *   while (*data != MARSH_EOS) {
             *       switch (*data) {
             *           ... some code ...
             *       }
             *   }
             */
            goto loop_end;
            default: /* not number, not string: invalid key */
            return 0;
        }
    }
loop_end:

    *datap = data;
    return 1;
}


    int
shelve_marshal(lua_State *L, char **data, int *bytes)
{
    size_t slen;
    size_t slen_aux;
    char ch;

    assert(L);
    assert(data);
    assert(bytes);
    assert(lua_gettop(L) > 0);

    switch (lua_type(L, -1)) {
        case LUA_TNUMBER: /* Encode a number. */
            slen = 1 + sizeof(double);
            {
                char __d[slen];
                __d[0] = MARSH_NUMBER;
                *((double*)(&__d[1])) = (double) lua_tonumber(L, -1);
                STOR(*data, *bytes, __d, slen);
            }
            break;
        case LUA_TBOOLEAN: /* Encode a boolean value. */
            ch = (lua_toboolean(L, -1)) ? MARSH_TRUE : MARSH_FALSE;
            STOR(*data, *bytes, &ch, 1);
            break;
        case LUA_TSTRING: /* Encode a string value. */
            slen = 1 + sizeof(size_t);
            {
                char __d[slen];
                const char *lstr = lua_tolstring (L, -1, &slen_aux);
                __d[0] = MARSH_STRING;
                *((size_t*)(&__d[1])) = slen_aux;
                STOR(*data, *bytes, __d, slen);
                STOR(*data, *bytes, lstr, slen_aux);
            }
            break;
        case LUA_TTABLE: /* Encode a table. */
            if (!marshal_table(L, data, bytes)) return 0;
        case LUA_TNIL: /* Just skip the 'nil' value. */
            break;
        case LUA_TFUNCTION:
        case LUA_TUSERDATA:
        case LUA_TLIGHTUSERDATA:
        default: /* Cannot handle userdatas or functions. */
            return 0;
    }

    ch = MARSH_EOS;
    STOR(*data, *bytes, &ch, 1);

    return 1;
}


int
marshal_table(lua_State *L, char **data, int *bytes)
{
    char ch = MARSH_TABLE;

    assert(L);
    assert(data);
    assert(bytes);
    assert(lua_type(L, -1) == LUA_TTABLE);

    STOR(*data, *bytes, &ch, 1);

    lua_pushnil(L);
    while (lua_next(L, -2)) {
        assert(lua_type(L, -3) == LUA_TTABLE);
        /*
         * - "key" is at index -2 and "value" at index -1
         * - "key" must be encoded first, so move it to the top
         * lua_insert() moves "value" down to -2 and shifts "key" up to -1
         * stack: key val tbl
         * index:  -1  -2  -3
         */
        lua_insert(L, -2);
        if (!shelve_marshal(L, data, bytes)) return 0; /* encode "key" */
        assert(lua_type(L, -3) == LUA_TTABLE);
        /*
         * Again, lua_insert() moves "key" down to -2 and shifted "value" up to -1.
         * stack: val key tbl
         * index:  -1  -2  -3
         */
        lua_insert(L, -2);
        if (!shelve_marshal(L, data, bytes)) return 0; /* encode "value" */
        assert(lua_type(L, -3) == LUA_TTABLE);
        /*
         * Pop "value", "key" remains in order to call lua_next() properly.
         * stack: key tbl
         * index:  -1  -2
         */
        lua_pop(L, 1);
        assert(lua_type(L, -2) == LUA_TTABLE);
    }

    assert(lua_type(L, -1) == LUA_TTABLE);

    /* Insert MARSH_EOS */
    ch = MARSH_EOT;
    STOR(*data, *bytes, &ch, 1);

    return 1;
}


int
l_shelve_marshal(lua_State *L)
{
    char *data = NULL;
    int sz  = 0;

    assert(L);

    if (shelve_marshal(L, &data, &sz)) {
        lua_pop(L, 1);
        lua_pushlstring(L, data, (size_t) sz);
        return 1;
    }

    lua_pop(L, 1);
    lua_pushnil(L);
    lua_pushstring(L, "Cannot encode data");
    return 2;
}


int
l_shelve_unmarshal(lua_State *L)
{
    const char *dta;

    assert(L);

    if (lua_gettop(L) != 1) {
        lua_pushstring(L, "one argument expected");
        lua_error(L);
        return 0;
    }

    if (!lua_isstring(L, -1)) {
        lua_pushstring(L, "argument must be of type string");
        lua_error(L);
        return 0;
    }

    dta = lua_tostring(L, -1);
    lua_pop(L, 1);

    if (!shelve_unmarshal(L, &dta)) {
        luaL_error(L, "bad format in encoded data");
    }

    return 1;
}



shelve.c
/*
 * Lua interface of the 'shelve' module.
 * Copyright (C) 2003-2016 Adrian Perez de Castro <aperez@igalia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

/* Deal with autoconf. */
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <lua.h>
#include <lauxlib.h>
#include <assert.h>
#include "defs.h"
#include "shelve.h"
#include "anydb.h"

#ifndef LUALIB_API
# define LUALIB_API API
#endif


/* Function protos */
int l_shelve_index(lua_State*);
int l_shelve_nindex(lua_State*);
int l_shelve_trv(lua_State*);
int l_shelve_gc(lua_State*);
int l_shelve_open(lua_State*);
int l_shelve_tostring(lua_State*);


/* Lua userdata type */
typedef struct shelve_file_t {
    anydb_t dbf;
    char *fname;
    int  rdonly;
} shelve_file;

/* Metatable items */


#define SHELVE_META_ITEMS 5
static luaL_Reg meta[] =
{
    { "__index",    l_shelve_index    },
    { "__newindex", l_shelve_nindex   },
    { "__call",     l_shelve_trv      },
    { "__gc",       l_shelve_gc       },
    { "__tostring", l_shelve_tostring },
};


LUALIB_API int
luaopen_shelve(lua_State *L)
{
    unsigned i;

    assert(L);

    /* This is more polite to loadmodule and luacheia */
    if (lua_gettop(L) == 0) {
        lua_pushstring(L, "shelve");
    } else {
        luaL_checktype(L, -1, LUA_TSTRING);
    }

    /* Create the namespace table. */
    lua_newtable(L);

    /* Create the metatable. */
    lua_pushstring(L, SHELVE_REGISTRY_KEY);
    lua_newtable(L);
    for (i=0 ; i<SHELVE_META_ITEMS ; i++) {
        lua_pushstring(L, meta[i].name);
        lua_pushcfunction(L, meta[i].func);
        lua_rawset(L, -3);
    }
    lua_settable(L, LUA_REGISTRYINDEX);

    /* Set the "open" function. */
    lua_pushstring(L, "open");
    lua_pushcfunction(L, l_shelve_open);
    lua_rawset(L, -3);

    /* Set the marshal/unmarshal functions. */
    lua_pushstring(L, "marshal");
    lua_pushcfunction(L, l_shelve_marshal);
    lua_rawset(L, -3);
    lua_pushstring(L, "unmarshal");
    lua_pushcfunction(L, l_shelve_unmarshal);
    lua_rawset(L, -3);

    return 1;
}


int
l_shelve_open(lua_State *L)
{
    int flags = ANYDB_WRITE;
    char *filename = NULL;
    shelve_file *udata = NULL;
    const char *rwmode = NULL;
    anydb_t dbh;
    int n;

    assert(L);

    n = lua_gettop(L);
    /* Check arguments. */
    if ((n != 1) && (n != 2)) {
        luaL_error(L, "function takes one or two arguments");
    }

    /* Check & get second argument (if needed). */
    if (n == 2) {
        rwmode = luaL_checkstring(L, -1);
        flags  = (*rwmode == 'r') ? ANYDB_READ : ANYDB_WRITE;
        lua_pop(L, 1);
    }

    /* Check & get DB file name. */
    filename = xstrdup(luaL_checkstring(L, -1));
    lua_pop(L, 1);

    /* Open the DB and remove filename from the stack. */
    if ( !(dbh = anydb_open(filename, flags)) ) {
        lua_pushnil(L);
        lua_pushstring(L, strerror(errno));
        return 2;
    }

    /* Configure returned userdata. */
    udata = (shelve_file*) lua_newuserdata(L, sizeof(shelve_file));
    udata->fname  = filename;
    udata->dbf    = dbh;
    udata->rdonly = (flags == ANYDB_READ);

    /* Associate metatable with userdata. */
    lua_pushstring(L, SHELVE_REGISTRY_KEY);
    lua_gettable(L, LUA_REGISTRYINDEX);
    lua_setmetatable(L, -2);
    lua_remove(L, -2);

    return 1;
}


int
l_shelve_index(lua_State *L)
{
    datum d, k;
    anydb_t *dbh;
    const char *datap;
    size_t slen_aux;

    assert(L);
    assert(lua_isuserdata(L, -2));

    dbh     = (anydb_t*) lua_touserdata(L, -2);
    k.dptr  = (char*) lua_tolstring(L, -1, &slen_aux);
    k.dsize = (int) slen_aux;

    d = anydb_fetch(*dbh, k);
    lua_pop(L, 2);

    if (!d.dptr) {
        lua_pushnil(L);
    }
    else {
        datap = d.dptr;

        if (!shelve_unmarshal(L, &datap)) {
            luaL_error(L, "bad format in encoded data");
        }
        xfree(d.dptr);
    }
    return 1;
}


int
l_shelve_nindex(lua_State *L)
{
    datum k, d = { NULL, 0 };
    shelve_file *udata;
    size_t slen_aux;

    assert(L);
    assert(lua_gettop(L) == 3);
    assert(lua_isuserdata(L, -3));

    udata   = (shelve_file*) lua_touserdata(L, -3);
    k.dptr  = (char*) lua_tolstring(L, -2, &slen_aux);
    k.dsize = (int) slen_aux;

    if (udata->rdonly) {
        luaL_error(L, "cannot modify read-only shelf datafile");
    }

    if (lua_isnil(L, -1)) {
        /* Remove key in database. */
        anydb_delete(udata->dbf, k);
        lua_pop(L, 3);
        return 0;
    }

    if (!shelve_marshal(L, &d.dptr, &d.dsize)) {
        luaL_error(L, "cannot encode data");
    }

    if (anydb_store(udata->dbf, k, d, ANYDB_REPLACE) != 0) {
        xfree(d.dptr);
        luaL_error(L, "cannot update item in data file");
    }
    xfree(d.dptr);

    lua_pop(L, 3);
    return 0;
}


int
l_shelve_trv(lua_State *L)
{
    anydb_t *dbh;
    datum k, tk;

    assert(L);
    assert(lua_gettop(L) == 1);
    assert(lua_isuserdata(L, -1));

    dbh = (anydb_t*) lua_touserdata(L, -1);
    lua_pop(L, 1);

    /*
     * A table containing all the keys is returned, all the keys
     * are assigned to 'true' boolean values. This is useful to
     * write Lua code similar to the following:
     *
     *    db = shelve.open("test.db")
     *    for key in db() do
     *        print("key: " .. key, "data: " .. db[key])
     *    end
     *
     */
    lua_newtable(L);
    for (k=anydb_firstkey(*dbh) ;
         k.dptr ;
         tk=k, k=anydb_nextkey(*dbh, k), xfree(tk.dptr))
    {
        lua_pushlstring(L, k.dptr, (size_t) k.dsize);
        lua_pushboolean(L, 1);
        lua_rawset(L, -3);
    }

    return 1;
}


int
l_shelve_gc(lua_State *L)
{
    shelve_file *udata;

    assert(L);
    assert(lua_gettop(L) == 1);
    assert(lua_isuserdata(L, -1));

    udata = (shelve_file*) lua_touserdata(L, -1);
    if (!udata->rdonly) anydb_reorganize(udata->dbf);
    anydb_close(udata->dbf);
    xfree(udata->fname);
    lua_pop(L, 1);
    return 0;
}


int
l_shelve_tostring(lua_State *L)
{
    shelve_file *udata;

    assert(L);
    assert(lua_gettop(L) == 1);
    assert(lua_isuserdata(L, -1));

    udata = (shelve_file*) lua_touserdata(L, -1);
    lua_pushfstring(L, "shelf (%s, %s)", udata->fname,
                    (udata->rdonly) ? "ro" : "rw");
    lua_remove(L, -2);
    return 1;
}

